pkg/promtail/targets/syslogtarget.go

```diff
@@ -0,0 +1,323 @@
+		Help:      "Total number of successful entries sent to the syslog target",
+	})
+	syslogParsingErrors = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "syslog_target_parsing_errors_total",
+		Help:      "Total number of parsing errors while receiving syslog messages",
+	})
+
+	defaultIdleTimeout = 120 * time.Second
+)
+
+// SyslogTarget listens to syslog messages.
+type SyslogTarget struct {
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrape.SyslogTargetConfig
+	relabelConfig []*relabel.Config
+
+	listener net.Listener
+	messages chan message
+
+	shutdown          chan struct{}
+	connectionsClosed *sync.WaitGroup
+}
+
+type message struct {
+	labels  model.LabelSet
+	message string
+}
+
+// NewSyslogTarget configures a new SyslogTarget.
+func NewSyslogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	config *scrape.SyslogTargetConfig,
+) (*SyslogTarget, error) {
+
+	t := &SyslogTarget{
+		logger:        logger,
+		handler:       handler,
+		config:        config,
+		relabelConfig: relabel,
+
+		shutdown:          make(chan struct{}),
+		connectionsClosed: new(sync.WaitGroup),
+	}
+
+	t.messages = make(chan message)
+	go t.messageSender()
+
+	err := t.run()
+	return t, err
+}
+
+func (t *SyslogTarget) run() error {
+	l, err := net.Listen("tcp", t.config.ListenAddress)
+	l = conntrack.NewListener(l, conntrack.TrackWithName("syslog_target/"+t.config.ListenAddress))
+	if err != nil {
+		return fmt.Errorf("error setting up syslog target %w", err)
+	}
+	t.listener = l
+	level.Info(t.logger).Log("msg", "syslog listening on address", "address", t.ListenAddress().String())
+
+	t.connectionsClosed.Add(1)
+	go t.acceptConnections()
+
+	return nil
+}
+
+func (t *SyslogTarget) acceptConnections() {
+	defer t.connectionsClosed.Done()
+
+	l := log.With(t.logger, "address", t.listener.Addr().String())
+	var backoff time.Duration
+
+	for {
+		c, err := t.listener.Accept()
+		if err != nil {
+			select {
+			case <-t.shutdown:
+				level.Info(l).Log("msg", "syslog server shutting down")
+				return
+			default:
+			}
+			if ne, ok := err.(net.Error); ok && ne.Temporary() {
+				if backoff == 0 {
+					backoff = 5 * time.Millisecond
+				} else {
+					backoff *= 2
+				}
+				if max := 1 * time.Second; backoff > max {
+					backoff = max
+				}
+				level.Warn(l).Log("msg", "failed to accept syslog connection", "err", err, "retry_in", backoff)
+				time.Sleep(backoff)
+				continue
+			}
+
+			level.Error(l).Log("msg", "failed to accept syslog connection. quiting", "err", err)
+			return
+		}
+
+		t.connectionsClosed.Add(1)
+		go t.handleConnection(c)
+	}
+}
+
+func (t *SyslogTarget) handleConnection(cn net.Conn) {
+	defer t.connectionsClosed.Done()
+
+	c := &idleTimeoutConn{cn, t.idleTimeout()}
+
+	done := make(chan struct{})
+	defer close(done)
+	go func() {
+		select {
+		case <-done:
+		case <-t.shutdown:
+		}
+		c.Close()
+	}()
+
```

Comment:
When using contexts, this would be replaced with

```
	handlerCtx, cancel := context.WithCancel(t.ctx)
	defer cancel()
	go func() {
		<-handlerCtx.Done()
		_ = c.Close()
	}()
```

(assuming `t.ctx` is the main context, that is cancelled in `Stop` method)